Add docker ps ancestor filter for image

vdemeester opened the pull request:
Makes it possible to filter containers by image _ancestor_, using `--filter=ancestor=busybox` or `--filter=ancestor=8c2e06607696`. This is related to #13365, and for the image filter, it's actually ask sometimes on irc 🐸.

The `ancestor` filter support the following _notation_:
- [x] image
- [x] image:tag
- [X] image@digest
- [x] short-id
- [x] full-id

Still to do :
- [X] Support multiple `--filter=ancestor`
- [X] Complete integration tests with short-id to make sure it works
- [X] Add integration tests with `image@digest`

Signed-off-by: Vincent Demeester vincent@sbr.pm

duglin commented:
what about containers that have that image the distant history?  e.g. the image in question was used to make another image and that's 2nd image is the one used to create the container.

vdemeester commented:
Ah :stuck_out_tongue_closed_eyes:, this would make it way more complicated to implement then :sweat_smile:, and I'm not sure how you would handle from a user point of view. 

```gherkin
Given that it was implement
And I have 3 images : base -> middle -> final
When I'm doing ``--filter=image=middle``
Then should it filter all containers based on middle and final, or just those with final ?
```
I think the simpler the better (just filter with the image used when starting the container). And I don't really see a use case for *distant history* — I might be wrong though.

icecrime commented:
I only see (potential) value is we walk the ancestry tree, the easy case I can pretty much grep :-( Not convinced.

vdemeester commented:
@icecrime On the other hand I'm not convinced on the ``grep`` part :blush: (although I like ``grep`` a lot :smiley_cat:). There is a few cases where I wouldn't find grep useful:

- Same or common part of the name shared by some images and some containers :

```bash
$ docker ps -a | grep busybox
5c33c99ac89f        busybox         […]                         toto                
5ba009755259        maven           […]                         busybox1            
2aff61195a5a        maven           […]                         busybox  
```

- The ability to *chain* it. Let's say I want to stop all container based on the busybox images : with the filter it would be ``docker ps -aq --filter=image=busybox | xargs docker stop``, without ``docker ps -a | grep busybox | awk '{ print $1 }' | xargs docker stop`` (but as seen in the previous point I could get false positive).

I do agree these are probably edge cases and might not happen a lot *in real life* or not that much but it felt so easy to implement.

I also do see *potential* on the "walk the ancestry tree" but I'm not sure how it would look in a user point of view. Question like : 

- Do I always want to walk the ancestry tree (and this, filtering on a base image would return all container running these or images based on it) or not ?
- If not, how would I do to specify it needs to walk the tree or not ? (like ``--filter=image=debian&children``)
- Is there a real use case for that ? I don't know but maybe.

duglin commented:
I only see the ancestry usecase as being interesting so no need for a flag to say "look at entire history", it should do it automagically.

thaJeztah commented:
I can see use cases for taking ancestry when filtering, but I really wonder if this is what I would expect if I filter `name=foo`. I think I would expect to only get containers started from `foo`. Perhaps another flag to include ancestry?

thaJeztah commented:
sorry `name=foo` should obviously be `image=foo` in my previous comment :-)

duglin commented:
the reason I'm not sure "image=foo" is that interesting is because there isn't much of a difference between:
```
docker run -ti ubuntu some-command
```
and
```
docker build -f Dockerfile-FROM-ubuntu+some-command ...
docker run ...
```

I guess it might be a little useful to have: `image=foo` and `ancestor=foo` but I wonder how quickly people will opt for the ancestor variant once they realize they **really** want to look for more than the immediate parent image.

thaJeztah commented:
> but I wonder how quickly people will opt for the ancestor variant once they realize they really want to look for more than the immediate parent image.

True, in many cases, either immediate parent or "uses" an image are what you're looking for. From a UX perspective though, I'm really against calling the "ancestor" filter "image", because it doesn't reflect the intent and will definitely lead to a lot of support questions / invalid bug reports.

We can opt for *not* implementing immediate parent (for now) and only implement ancestor search once people start asking for it. Perhaps an easier name for the filter, would `"inherits"` be a good name? `--filter inherits=ubuntu:14.04`

To the filtering itself, what kind of searches should be supported? A quick list;

- `image`
- `image:tag`
- `image:tag@digest`
- `short-id`
- `full-id`

cpuguy83 commented:
I definitely +1 to more filters.
Grep is nice, but not particularly nice in this case.

tiborvass commented:
*Collective review*

@duglin @calavera @tonistiigi 

We would prefer the tree-traversal implementation instead of just filtering with a direct parent image name.

Because in the inspect output, Image refers to the direct parent image, we should not call the filter `image`, and we suggest `ancestor` instead (less ambiguous).

As for answering https://github.com/docker/docker/pull/14570#issuecomment-121414249, we should support all of the input forms. They will have to be mapped to a full image ID anyways.

Let's leave the "direct-parent-only" implementation for another time, as we seem to agree the tree-traversal is more useful.

thaJeztah commented:
SGTM, thanks!

vdemeester commented:
SGTM too, I'll work on that then :wink:

vdemeester commented:
Updated the PR with an implementation of the `ancestor` filter :wink:.

calavera commented on daemon/list.go:
```diff
@@ -61,6 +64,21 @@ func (daemon *Daemon) Containers(config *ContainersConfig) ([]*types.Container,
+			image, err := daemon.Repositories().LookupImage(ancestor)
+			if err == nil {
+				// Then walk down the graph and put the imageIds in imagesFilter
+				byParents := daemon.Graph().ByParent()
```
do we need to call `ByParent` in each iteration?

vdemeester replied:
Oh no, I'm pretty sure we don't, you're right :sweat_smile:. I'll fix that.

calavera commented on daemon/list.go:
```diff
@@ -203,3 +236,15 @@ func (daemon *Daemon) Containers(config *ContainersConfig) ([]*types.Container,
 }
+
+func populateImageFilterByParents(imageID string, byParents map[string][]*image.Image) []string {
+	var ids []string
```
`ids := []string{imageID}`?

vdemeester replied:
ah yeah :blush:

calavera commented on integration-cli/docker_cli_ps_test.go:
```diff
@@ -279,6 +282,118 @@ func (s *DockerSuite) TestPsListContainersFilterName(c *check.C) {
+		for i := range expectedIDs {
+			if actualIDs[i] != expectedIDs[i] {
+				c.Logf("%s, %s", actualIDs[i], expectedIDs[i])
+				same = false
```
break here? I don't think you need to keep iterating the slice.

vdemeester replied:
Oh right, we don't need to keep iterating there, thanks ! :sweat_smile:

calavera replied:
thanks

calavera commented on daemon/list.go:
```diff
@@ -61,6 +64,21 @@ func (daemon *Daemon) Containers(config *ContainersConfig) ([]*types.Container,
+		for _, ancestor := range ancestors {
+			// First, get the imageId of the ancestor filter (yay)
+			image, err := daemon.Repositories().LookupImage(ancestor)
+			if err == nil {
```
what happens when there is an error? is it okay to just ignore it like we're doing now or we should fail?

duglin replied:
From my scan of the code, I think its ok to ignore it.  I think we'd only get an error if there's something wrong with the graph, which shouldn't really prevent this command from running. We may get less results than expected, but that's better than just stopping.  Perhaps logging it (in the LookupImage() ) func would be good though so an admin watching the logs can investigate.

vdemeester replied:
Oh right, we should log it in case of error, but I agree with @duglin, I thinks its safe to go on (and not fail right away).

duglin commented on daemon/list.go:
```diff
@@ -61,6 +64,21 @@ func (daemon *Daemon) Containers(config *ContainersConfig) ([]*types.Container,
+			image, err := daemon.Repositories().LookupImage(ancestor)
+			if err == nil {
+				// Then walk down the graph and put the imageIds in imagesFilter
+				imagesFilter = append(imagesFilter, populateImageFilterByParents(image.ID, byParents)...)
```
If one of the ancestors happens to be a child of a previous 'ancestor' then we'll end up doing this recursive stuff twice for the same image ID (I think).  If so, we may need to add an if-statement before this line to check to see if 'ancestor' is already in the 'imagesFilter' list.  Just to optimize things a bit. But please double-check my thinking.

vdemeester replied:
@duglin right :smile_cat:. And if it was a map it would be really simple to check. Let's try that :wink:.

duglin commented on daemon/list.go:
```diff
@@ -131,12 +149,27 @@ func (daemon *Daemon) Containers(config *ContainersConfig) ([]*types.Container,
+
+		if ancestorFilter {
+			shouldSkip := true
+			for _, imageIDtoMatch := range imagesFilter {
```
If 'imagesFilter' where a map then I think you might be able to replace this for-loop with a hash lookup.

duglin commented:
I think we need doc updates.

duglin commented on integration-cli/docker_cli_by_digest_test.go:
```diff
@@ -387,6 +387,35 @@ func (s *DockerRegistrySuite) TestListImagesWithDigests(c *check.C) {
+	out, _ := dockerCmd(c, "run", "-d", imageReference, "echo", "hello")
+	expectedID := strings.TrimSpace(out)
+
+	// Imvalid imageReference
```
Imvalid -> Invalid    m vs n

vdemeester replied:
ah :dizzy_face:

duglin commented on integration-cli/docker_cli_ps_test.go:
```diff
@@ -279,6 +282,119 @@ func (s *DockerSuite) TestPsListContainersFilterName(c *check.C) {
+	imageID1, err := buildImage(imageName1,
+		`FROM busybox
+		 LABEL match me 1`, true)
+	if err != nil {
```
Just for consistency with the direction of other PRs, perhaps use:
```
c.Assert(err, check.IsNil)
```

vdemeester replied:
Ah well yeah, my bad :sweat_smile: ..

vdemeester commented:
@duglin took care of your comments. Implementation with a map feels easier to read (less for loops :smile_cat:).

I was waiting for a few reviews before updating docs :stuck_out_tongue_closed_eyes:, that will be my next task :wink:.

calavera commented on daemon/list.go:
```diff
@@ -28,13 +30,15 @@ type ContainersConfig struct {
+		psFilters      filters.Args
+		filtExited     []int
 	)
+	imagesFilter := map[string]struct{}{}
```
if you use `map[string]bool` you can simplify the conditionals to something like `if imagesFilter[imageID] {` and remove those extra `ok`

vdemeester replied:
@calavera Yeah I was wondering about that : "empty struct + extra `ok`" or a "`bool` and nothing else". The later takes a little bit more memory but probably not enough to not do it ; I think.

vdemeester committed:
Add 'ancestor' ps filter for image

Makes it possible to filter containers by image, using
--filter=ancestor=busybox and get all the container running busybox
image and image based on busybox (to the bottom).

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

calavera commented:
LGTM

duglin commented on integration-cli/docker_cli_by_digest_test.go:
```diff
@@ -387,6 +387,33 @@ func (s *DockerRegistrySuite) TestListImagesWithDigests(c *check.C) {
+	out, _ = dockerCmd(c, "ps", "-a", "-q", "--no-trunc", "--filter=ancestor="+imageReference)
+	if strings.TrimSpace(out) != expectedID {
+		c.Fatalf("Expected filter container for %s ancestor filter to be %v, got %v", imageReference, expectedID, strings.TrimSpace(out))
+	}
```
In the other test (below) you check for ancestors and immediate parent by ID, name, etc... but not digest. Here you're just checking for immediate parent by digest.  I would hope checking for ancestor by digest would just work too, since its just another way to reference an image, but is it worth adding at least one check for "ancestor by digest" instead of just immediate parent?

vdemeester replied:
Ah right, building an image from the `imageReference` and make sure we get the *digested* one and its descendant.

duglin commented:
looks good, just one question.

vdemeester commented:
Oh I though I pushed the docs (which are quite small) but didn't.. :sweat_smile:. /cc @thaJeztah

thaJeztah commented on docs/reference/commandline/ps.md:
```diff
@@ -50,6 +50,8 @@ The currently supported filters are:
 * name (container's name)
 * exited (int - the code of exited containers. Only useful with `--all`)
 * status (created|restarting|running|paused|exited)
+* ancestor (image id) - filters containers that were created from the
```
we probably should avoid using "image id" here, because the filter also accepts "image name", right?

Will need some examples, to illustrate/describe the possible options

vdemeester replied:
@thaJeztah oh right, we should probably just add `image` (so it stays general) there and add more example in a `Filtering` chapter (as discussed a bit on IRC :wink:)

vdemeester commented:
Updated with to take care of @duglin and @thaJeztah comments, I think :sweat_smile:.

thaJeztah commented:
docs LGTM, as discussed with @vdemeester,  examples will be added in a filtering chapter/section https://github.com/docker/docker/pull/14570#discussion_r37805119

duglin commented on integration-cli/docker_cli_by_digest_test.go:
```diff
@@ -387,6 +387,43 @@ func (s *DockerRegistrySuite) TestListImagesWithDigests(c *check.C) {
+	// pull from the registry using the <name>@<digest> reference
+	dockerCmd(c, "pull", imageReference)
+
+	// build a container from it
```
nit, s/container/image/

vdemeester replied:
ah "*oups*" :sweat_smile:

duglin commented:
@vdemeester LGTM

cpuguy83 commented:
Moving to docs review.

thaJeztah commented:
LGTM per my previous comment https://github.com/docker/docker/pull/14570#issuecomment-134394517

thaJeztah commented:
ping @moxiegirl, looks like we merged the "follow up" before the one it's "following up" :smile:

vdemeester commented:
@thaJeztah yay :sweat_smile: :stuck_out_tongue_closed_eyes:

moxiegirl commented:
This is why I should leave the merging to others (@thaJeztah) ...
![](http://gifs.joelglovier.com/fail/truck-pool.gif)

thaJeztah commented:
@moxiegirl counting that as a LGTM, merging!

vdemeester commented:
:tada: :wink:
